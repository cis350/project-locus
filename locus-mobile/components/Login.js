import React, { useState } from 'react';
import {
  View, TouchableHighlight, Text, StyleSheet, TextInput, Alert,
} from 'react-native';
import { login, getUser } from '../modules/api';

export default function Login({ navigation }) {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');

  async function handleLogin() {
    const loginStatus = await login(email, password);
    if (loginStatus === 404) {
      Alert.alert('Invalid Username');
    } else if (loginStatus === 400) {
      Alert.alert('Invalid Password');
    } else if (loginStatus === 403) {
      Alert.alert('Account Locked, Try Again Later');
    } else {
      const user = (await getUser(email)).result;
      if (!user) return;
      navigation.navigate('TabNavigation', { user });
    }
  }

  return (
    <View style={styles.container}>
      <Text style={{ textAlign: 'center', fontSize: 40 }}>Login</Text>
      <View style={styles.formContainer}>
        <TextInput
          style={styles.input}
          placeholder="Email"
          placeholderTextColor="grey"
          onChangeText={setEmail}
          value={email}
        />
        <TextInput
          style={styles.input}
          placeholder="Password"
          placeholderTextColor="grey"
          onChangeText={setPassword}
          value={password}
          // eslint-disable-next-line react/jsx-boolean-value
          secureTextEntry={true}
        />
      </View>
      <TouchableHighlight style={styles.button} underlayColor="#33E86F" onPress={() => handleLogin()}>
        <Text style={{ fontSize: 30 }}>Login</Text>
      </TouchableHighlight>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flexDirection: 'column',
    alignItems: 'center',
    paddingTop: 100,
  },
  formContainer: {
    flexDirection: 'column',
    justifyContent: 'space-between',
    backgroundColor: '#B5E48C',
    paddingVertical: 30,
    paddingHorizontal: 20,
    borderRadius: 10,
    marginVertical: 30,
  },
  input: {
    backgroundColor: 'white',
    borderColor: 'gray',
    width: 250,
    borderWidth: 1,
    borderRadius: 10,
    padding: 10,
    marginVertical: 5,
  },
  button: {
    backgroundColor: '#6A9B72',
    borderRadius: 10,
    paddingVertical: 10,
    width: 150,
    justifyContent: 'center',
    alignItems: 'center',
    marginHorizontal: 10,
  },
});
