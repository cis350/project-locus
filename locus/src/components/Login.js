import React, { useState } from 'react';
import { useNavigate } from 'react-router-dom';
import {
  Button,
  Card, Form,
  Alert,
} from 'react-bootstrap';
// import { verifyLogInInfo, getUserUniqueId } from '../modules/storage';

const Login = function LoginComponent({ setIsLoggedIn, setUserEmail }) {
  const [logInEmail, setLogInEmail] = useState('');
  const [logInPassword, setLogInPassword] = useState('');

  const [logInFieldEmpty, setLogInFieldEmpty] = useState(false);
  const [logInInfoInvalid, setLogInInfoInvalid] = useState(false);

  const navigate = useNavigate();

  let logInData;

  function setLogInData(data) {
    logInData = data;
  }

  // handles redirecting to "/home"
  function onLogIn(path) {
    navigate(path);
    setIsLoggedIn(true);
    setUserEmail(logInEmail);
  }

  const updateEmail = (e) => {
    setLogInEmail(e.target.value);
  };

  const updatePassword = (e) => {
    setLogInPassword(e.target.value);
  };

  const processUserInputs = () => {
    if (logInEmail === '' || logInPassword === '') {
      setLogInFieldEmpty(true);
      setLogInInfoInvalid(false);
    }

    fetch('/login').then((res) => res.json()).then((data) => setLogInData(data));

    if (logInData.status === 200) {
      setLogInFieldEmpty(false);
      setLogInInfoInvalid(false);
      onLogIn(`/home/${logInData.userId}`);
    } else {
      setLogInFieldEmpty(false);
      setLogInInfoInvalid(true);
    }

    // else if (!verifyLogInInfo(logInEmail, logInPassword)) {
    //   setLogInFieldEmpty(false);
    //   setLogInInfoInvalid(true);
    // } else {
    //   const uniqueId = getUserUniqueId(logInEmail);
    //   setLogInFieldEmpty(false);
    //   setLogInInfoInvalid(false);
    //   onLogIn(`/home/${uniqueId}`);
    // }
  };

  const errorMsgEmptyFields = (() => (
    // referenced https://react-bootstrap.github.io/components/alerts/
    <Alert variant="danger" style={{ width: '23rem', margin: 'auto', marginTop: '10px' }} className="text-center">
      Please enter both your email and password.
    </Alert>
  ));

  const errorMsgLogInInfoInvalid = (() => (
    // referenced https://react-bootstrap.github.io/components/alerts/
    <Alert variant="danger" style={{ width: '23rem', margin: 'auto', marginTop: '10px' }} className="text-center">
      Email or/and password invalid.
    </Alert>
  ));

  return (
    <div className="container" style={{ position: 'relative', padding: '120px' }}>
      <h1 className="text-center">Log-in</h1>
      {logInFieldEmpty && errorMsgEmptyFields()}
      {logInInfoInvalid && errorMsgLogInInfoInvalid()}
      <Card style={{
        width: '23rem',
        margin: 'auto',
        marginTop: '20px',
        borderRadius: '10px',
        backgroundColor: '#B5E48C',
        borderColor: '#B5E48C',
      }}
      >
        <Card.Body>
          {/* referenced https://react-bootstrap.github.io/forms/overview/ */}
          <Form>
            <Form.Group className="mb-3" controlId="formBasicEmail">
              <Form.Label>Email</Form.Label>
              <Form.Control style={{ height: '25px' }} type="email" onChange={(e) => updateEmail(e)} />
              <Form.Label>Password</Form.Label>
              <Form.Control style={{ height: '25px' }} type="password" onChange={(e) => updatePassword(e)} />
            </Form.Group>
          </Form>
        </Card.Body>
      </Card>
      <div className="navbar-brand" style={{ display: 'flex', justifyContent: 'center', marginTop: '30px' }}>
        <Button
          style={{
            backgroundColor: '#6A9B72',
            width: '120px',
            height: '50px',
            fontWeight: 'bold',
            fontSize: '25px',
            color: 'black',
            borderColor: '#6A9B72',
          }}
          onClick={() => processUserInputs()}
        >
          Log-in
        </Button>
      </div>
    </div>
  );
};

export default Login;
