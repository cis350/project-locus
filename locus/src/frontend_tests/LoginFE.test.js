/**
* @jest-environment jsdom
*/
import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import Login from '../components/Login';

const axios = require('axios');
const MockAdapter = require('axios-mock-adapter');

const domain = 'http://localhost:3306';

// mock useNavigate: https://github.com/remix-run/react-router/issues/7811
const mockedNavigator = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedNavigator,
}));

let mock;
beforeAll(() => {
  mock = new MockAdapter(axios);
});
afterEach(() => {
  mock.reset();
});

describe('Initial Login page displays', () => {
  test('Displays Log-in text', () => {
    render(<Login />);
    const textElement = screen.getAllByText(/Log-in/i);
    expect(textElement[0]).toBeInTheDocument();
  });
  test('Displays Email', () => {
    render(<Login />);
    const textElement = screen.getByText(/Email/i);
    expect(textElement).toBeInTheDocument();
  });
  test('Displays Password', () => {
    render(<Login />);
    const textElement = screen.getByLabelText(/Password/i);
    expect(textElement).toBeInTheDocument();
  });
  test('Displays log-in button', () => {
    render(<Login />);
    const button = screen.getByRole('button', { name: /Log-in/i });
    expect(button).toBeInTheDocument();
  });
  test('Displays reset password button', () => {
    render(<Login />);
    const button = screen.getByRole('button', { name: /reset/i });
    expect(button).toBeInTheDocument();
  });
});

describe('Alert message tests', () => {
  test('empty fields', () => {
    render(<Login />);
    userEvent.click(screen.getByRole('button', { name: /Log-in/i }));
    const alertElement = screen.getByText('Please enter both your email and password.');
    expect(alertElement).toBeInTheDocument();
  });
  test('password field empty', () => {
    render(<Login />);
    userEvent.type(screen.getByTestId('email-input'), 'test@gmail.com');
    userEvent.click(screen.getByRole('button', { name: /Log-in/i }));
    const alertElement = screen.getByText('Please enter both your email and password.');
    expect(alertElement).toBeInTheDocument();
  });
  test('email field empty', () => {
    render(<Login />);
    userEvent.type(screen.getByTestId('password-input'), 'pass');
    userEvent.click(screen.getByRole('button', { name: /Log-in/i }));
    const alertElement = screen.getByText('Please enter both your email and password.');
    expect(alertElement).toBeInTheDocument();
    userEvent.click(screen.getByRole('button', { name: /reset/i }));
  });
  // test('invalid login', () => {
  //   render(<Login />);
  //   userEvent.type(screen.getByTestId('email-input'), 'test@gmail.com');
  //   userEvent.type(screen.getByTestId('password-input'), 'pass');
  //   userEvent.click(screen.getByRole('button', { name: /Log-in/i }));
  //   const alertElement = screen.getByText('Email or/and password invalid.');
  //   expect(alertElement).toBeInTheDocument();
  // });
});

describe('Login Button Tests', () => {
  test('Click Login with 200', () => {
    render(<Login />);
    userEvent.type(screen.getByTestId('email-input'), 'email');
    userEvent.type(screen.getByTestId('password-input'), 'pass');
    // mock.onPost(`${domain}/login`).reply(200);
    userEvent.click(screen.getByRole('button', { name: /Log-in/i }));
  });
  // test('Click Login with 400', () => {
  //   render(<Login />);
  //   userEvent.type(screen.getByTestId('email-input'), 'email');
  //   userEvent.type(screen.getByTestId('password-input'), 'pass');
  //   mock.onPost(`${domain}/login`).reply(400);
  //   userEvent.click(screen.getByRole('button', { name: /Log-in/i }));
  // });
  // test('Click Login with 403', () => {
  //   render(<Login />);
  //   userEvent.type(screen.getByTestId('email-input'), 'email');
  //   userEvent.type(screen.getByTestId('password-input'), 'pass');
  //   mock.onPost(`${domain}/login`).reply(403);
  //   userEvent.click(screen.getByRole('button', { name: /Log-in/i }));
  // });
  // test('Click Login with 404', () => {
  //   render(<Login />);
  //   userEvent.type(screen.getByTestId('email-input'), 'email');
  //   userEvent.type(screen.getByTestId('password-input'), 'pass');
  //   mock.onPost(`${domain}/login`).reply(404);
  //   userEvent.click(screen.getByRole('button', { name: /Log-in/i }));
  // });
});
