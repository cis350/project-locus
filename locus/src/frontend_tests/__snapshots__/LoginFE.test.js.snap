// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Test 1`] = `
<div
  className="container"
  style={
    Object {
      "padding": "120px",
      "position": "relative",
    }
  }
>
  <h1
    className="text-center"
  >
    Log-in
  </h1>
  <div
    className="card"
    style={
      Object {
        "backgroundColor": "#B5E48C",
        "borderColor": "#B5E48C",
        "borderRadius": "10px",
        "margin": "auto",
        "marginTop": "20px",
        "width": "23rem",
      }
    }
  >
    <div
      className="card-body"
    >
      <form
        className=""
      >
        <div
          className="mb-3"
        >
          <label
            className="form-label"
            htmlFor="formBasicEmail"
          >
            Email
          </label>
          <input
            className="form-control"
            id="formBasicEmail"
            onChange={[Function]}
            style={
              Object {
                "height": "25px",
              }
            }
            type="email"
          />
          <label
            className="form-label"
            htmlFor="formBasicEmail"
          >
            Password
          </label>
          <input
            className="form-control"
            id="formBasicEmail"
            onChange={[Function]}
            style={
              Object {
                "height": "25px",
              }
            }
            type="password"
          />
        </div>
      </form>
    </div>
  </div>
  <div
    className="navbar-brand"
    style={
      Object {
        "display": "flex",
        "justifyContent": "center",
        "marginTop": "30px",
      }
    }
  >
    <button
      className="btn btn-primary"
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "#6A9B72",
          "borderColor": "#6A9B72",
          "color": "black",
          "fontSize": "25px",
          "fontWeight": "bold",
          "height": "50px",
          "width": "120px",
        }
      }
      type="button"
    >
      Log-in
    </button>
  </div>
</div>
`;
